freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/flow/FreeIpaCloudStorageValidationService.java

```diff
@@ -0,0 +1,80 @@
+package com.sequenceiq.freeipa.service.freeipa.flow;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.orchestrator.exception.CloudbreakOrchestratorException;
+import com.sequenceiq.cloudbreak.orchestrator.exception.CloudbreakOrchestratorFailedException;
+import com.sequenceiq.cloudbreak.orchestrator.host.HostOrchestrator;
+import com.sequenceiq.cloudbreak.orchestrator.model.GatewayConfig;
+import com.sequenceiq.cloudbreak.orchestrator.model.Node;
+import com.sequenceiq.cloudbreak.orchestrator.model.SaltConfig;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.orchestrator.StackBasedExitCriteriaModel;
+import com.sequenceiq.freeipa.service.GatewayConfigService;
+import com.sequenceiq.freeipa.service.freeipa.config.FreeIpaSaltConfigService;
+import com.sequenceiq.freeipa.service.stack.StackService;
+
+@Service
+public class FreeIpaCloudStorageValidationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaCloudStorageValidationService.class);
+
+    @Inject
+    private HostOrchestrator hostOrchestrator;
+
+    @Inject
+    private GatewayConfigService gatewayConfigService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaSaltConfigService freeIpaSaltConfigService;
+
+    public void validate(Long stackId) throws CloudbreakOrchestratorException {
+        Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+        Set<InstanceMetaData> instanceMetaDatas = stack.getNotDeletedInstanceMetaDataSet();
+        List<GatewayConfig> gatewayConfigs = gatewayConfigService.getGatewayConfigs(stack, instanceMetaDatas);
+        Set<Node> allNodes = mapInstancesToNodes(instanceMetaDatas);
+
+        createAndPushPillarsToNodes(stackId, stack, gatewayConfigs, allNodes);
```

Comment:
just wondering if we could avoid pushing the same pillars twice. maybe a new flowstep would be necessary. I'm concerned because creating the pillars includes multiple interservice calls also, so it could take a bit of time.